### ISSUE TITLE
What is the best way to display the log?

### ISSUE THREAD

**Issue by wrux** [spatie/laravel-activitylog#360]
In my user profile I am trying to display the log information. Is there an easy way in Laravel that I can display completely different content based on the subject model? 

So if the user created a Product then it shows the information on the product then if they edit the product it just shows the fields they updated for example.

At the minute i'm just doing a `@foreach($activity as $log)` in the template view and showing the user/desctiption

**Comment by wrux** [comment]
Also this is off topic but I cant see it in the documentation either. Is it possible to set a log_name value when using the `activity()` function?

**Comment by freekmurze** [comment]
> At the minute i'm just doing a @foreach($activity as $log) in the template view and showing the user/desctiption

That would be the easy way to go. The `Activity` model is just an Eloquent model. You can do anything with it you want. The package doesn't have an opinion on how the activity should be rendered. You need to take care of that yourself.

> Also this is off topic but I cant see it in the documentation either. Is it possible to set a log_name value when using the activity() function?

It's in the docs: https://docs.spatie.be/laravel-activitylog/v2/advanced-usage/using-multiple-logs#specifying-a-log

**Comment by Gummibeer** [comment]
If you want to render an Activity based in related models the first idea I get is Wordpress style - you have multiple view partials for all cases.

```
/views/activities/...
.../default.blade.php
.../product.blade.php
.../comment.blade.php
```

These are usable in the "timeline" with a `switch-case`.

```php
switch(get_class($activity->subject){
 case 'App\Product':
  // render activity partials
 ...
}
```

Sorry, just mobile so no complete Code - but the idea should be clear!?
I think that this is the cleanest was to render also big parts totally different.

**Comment by wrux** [comment]
Hi this is what I am trying to do. 

I have stored `$logTemplate = 'product';` in my models and then i my loop I was tryning to include a file based on the $logTemplate variable but I can't seem to get the $logTemplate var

**Comment by pbgneff** [comment]
In the same vein of this issue. I have my logger set up to only log the dirty attributes from updating models. Because of that the properties in the db are grouped like so:

```json
{  
   "attributes":{  
      "attribute":"updated value"
   },
   "old":{  
      "attribute":"old value"
   }
}
```

I'm having issues displaying this data dynamically in a human friendly way, since I have no idea what the attribute is. I've resolved to some pretty ugly code of looping through the properties and grabbing the first `$key` and then concatenating the `$values`. Does anyone have a better idea? 

Alternatively, is there a way to customize the way the properties are logged for all model updates (as in globally or something)?

**Comment by pbgneff** [comment]
That's what I was doing, it's ugly. I was thinking of submitting a PR with a new json structure when using `logOnlyDirty`:

```json
{  
   "attribute":"attribute_name",
   "values":{  
      "from":"old value",
      "to":"new value"
   }
}
```

**Comment by Gummibeer** [comment]
Not as a PR please. To have different structure depending on your configuration will end up in a pain if you want to see all activities. You also would have to migrate all old records and so on.

But what's your problem with the current structure? You are able to get all data without any black-magic.

**Comment by pbgneff** [comment]
In terms of migrating old records I see your point, because that would be a big issue. Although I have a couple ideas for how to resolve that.

In terms of the the point. I saw some people posting questions regarding how to display the data, and so I'd be curious if this json structure would be more appealing moving forward.

My use case is I'd like to use the activity log to display recent user activity. I could go macro and just have a model specific sentence for updates (User was updated) but it would be great to go a little deeper (Updated first name from "John" to "Mike" for User).

So this is about my specific use case, but I'd be curious to see if when people use `logOnlyDirty` they'd be interested in more specific properties.

Or allow overriding the `withProperties` for model logging.

As of now I'm going to go with the loop alternative you mentioned, but as you mentioned it's not recommended.
